pennylane_lightning/core/src/simulators/lightning_gpu/StateVectorCudaBase.hpp

```diff
@@ -0,0 +1,292 @@
+// Copyright 2022-2023 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file StateVectorCudaBase.hpp
+ */
+#pragma once
+
+#include "cuda.h"
+#include <cuda_runtime_api.h> // cudaMalloc, cudaMemcpy, etc.
+#include <memory>
+#include <type_traits>
+#include <unordered_map>
+#include <unordered_set>
+
+#include "DataBuffer.hpp"
+#include "DevTag.hpp"
+#include "Error.hpp"
+#include "StateVectorBase.hpp"
+#include "cuda_helpers.hpp"
+
+/// @cond DEV
+namespace {
+namespace cuUtil = Pennylane::LightningGPU::Util;
+// using namespace Pennylane::Util;
+} // namespace
+/// @endcond
+
+namespace Pennylane::LightningGPU {
+// declarations of external functions (defined in initSV.cu).
+extern void setBasisState_CUDA(cuComplex *sv, cuComplex &value,
+                               const size_t index, bool async,
+                               cudaStream_t stream_id);
+extern void setBasisState_CUDA(cuDoubleComplex *sv, cuDoubleComplex &value,
+                               const size_t index, bool async,
+                               cudaStream_t stream_id);
+/**
+ * @brief CRTP-enabled base class for CUDA-capable state-vector simulators.
+ *
+ * @tparam Precision Floating point precision.
+ * @tparam Derived Derived class to instantiate using CRTP.
+ */
+template <class Precision, class Derived>
+class StateVectorCudaBase : public StateVectorBase<Precision, Derived> {
+  private:
+    using BaseType = StateVectorBase<Precision, Derived>;
+
+  public:
+    /// Alias for accessing the class templated precision type
+    using scalar_type_t = Precision;
+
+    /// Alias for complex floating point datatype supported by class
+    using CFP_t = decltype(cuUtil::getCudaType(Precision{}));
```

Comment:
Can we rename this here and all the uses of it below?

Reply:
As disscussed, let's update `CFP_t` later in a follow-up PR.